### PR TITLE
TDI-43870 : FTP

### PR DESCRIPTION
**Pull request by rlecomte-talend**
https://jira.talendforge.org/browse/TDI-43870

### CODE REVIEW

File: ftp/pom.xml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        <artifactId>connectors-se</artifactId>
+        <version>1.8.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>ftp</artifactId>
+
+    <name>Components :: FTP</name>
+
+    <properties>
+        <testcontainers.version>1.12.3</testcontainers.version>
```
Comment (undx) [inline ftp/pom.xml]: As the number of components using testContainers is growing, we should now put this in parent pom I think.

File: ftp/pom.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.talend.components</groupId>
+            <artifactId>connectors-test-bom</artifactId>
+            <type>pom</type>
+        </dependency>
+        <dependency>
+            <groupId>commons-net</groupId>
+            <artifactId>commons-net</artifactId>
+            <version>3.6</version>
```
Comment (undx) [inline ftp/pom.xml]: put version in properties.

File: ftp/pom.xml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.jcraft</groupId>
+            <artifactId>jsch</artifactId>
+            <version>0.1.55</version>
```
Comment (undx) [inline ftp/pom.xml]: put version in properties.

File: ftp/pom.xml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>${testcontainers.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockftpserver</groupId>
+            <artifactId>MockFtpServer</artifactId>
+            <version>2.7.1</version>
```
Comment (undx) [inline ftp/pom.xml]: put version in properties.

File: ftp/src/main/java/org/talend/components/ftp/dataset/FTPDataSet.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    public ContentFormat getFormatConfiguration() {
+        try {
+            Field f = FTPDataSet.class.getDeclaredField(format.getConfigName());
+            return (ContentFormat) f.get(this);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            log.error(e.getMessage(), e);
+            return null;
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/dataset/FTPDataSet.java]: I don't get the goal of setting configName with PROTECTED level and using getDeclareField in this class...

File: ftp/src/main/java/org/talend/components/ftp/datastore/FTPDataStore.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+@Checkable(FTPService.ACTION_HEALTH_CHECK)
+@GridLayout(names = GridLayout.FormType.MAIN, value = { @GridLayout.Row("fileProtocol"), @GridLayout.Row("host"),
+        @GridLayout.Row("port"), @GridLayout.Row("useCredentials"), @GridLayout.Row({ "username", "password" }) })
+@GridLayout(names = GridLayout.FormType.ADVANCED, value = { @GridLayout.Row({ "trustType", "protocol" }),
+        @GridLayout.Row("keepAliveTimeout"), @GridLayout.Row("keepAliveReplyTimeout") })
+@Documentation("FTP connection Properties")
+public class FTPDataStore implements Serializable {
+
+    @Option
+    @Documentation("Type of protocol to use to connect")
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/datastore/FTPDataStore.java]: forget the dot.

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            // Must create new file
+            String path = fileBaseName + "_" + fileIndex++ + "." + configuration.getDataSet().getFormat().getExtension();
+            log.debug("Creating remote file " + path);
+            currentStream = new SizeAwareOutputStream(getFtpClient().storeFileStream(path));
+            ContentFormat contentFormat = configuration.getDataSet().getFormatConfiguration();
+            recordWriter = recordIORepository.findWriter(contentFormat.getClass()).getWriter(() -> currentStream, contentFormat);
+            try {
+                recordWriter.init(contentFormat);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java]: use log instead

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            }
+        }
+
+    }
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java]: Here, when there's an exception you just log it, won't break flow?

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutputConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            @ActiveIf(target = "limitBy", value = "BOTH") })
+    @Documentation("Max number of records per file.")
+    private int recordsLimit = 1000;
+
+    @Option
+    @DefaultValue("1000")
+    @ActiveIfs(operator = ActiveIfs.Operator.OR, value = { @ActiveIf(target = "limitBy", value = "SIZE"),
+            @ActiveIf(target = "limitBy", value = "BOTH") })
+    @Documentation("Max file size.")
+    private int sizeLimit = 10;
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/output/FTPOutputConfiguration.java]: You put @DefaultValue to 1000 and initialize member to 10.

File: ftp/src/main/java/org/talend/components/ftp/service/ftpclient/ApacheFTPClient.java [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+@Slf4j
+/**
+ * FTP client for FTP and FTPS
+ */
+public class ApacheFTPClient extends GenericFTPClient {
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/service/ftpclient/ApacheFTPClient.java]: Strange name...

File: ftp/src/main/java/org/talend/components/ftp/service/ftpclient/FTPClientFactory.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        // ftpClient = ApacheFTPClient.createFTPS(datastore);
+        // break;
+        // case SFTP:
+        // ftpClient = JschFTPSClient.create(datastore);
+        // break;
+        default:
+            throw new FTPConnectorException("Unknown File protocol : " + datastore.getFileProtocol());
+        }
+
+        ftpClient.setI18n(i18n);
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/service/ftpclient/FTPClientFactory.java]: injection won't work?

File: ftp/src/main/java/org/talend/components/ftp/service/ftpclient/JschFTPSClient.java [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.List;
+import java.util.Vector;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+@Slf4j
+/**
+ * FTP client for SFTP connection
+ */
+public class JschFTPSClient extends GenericFTPClient {
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/service/ftpclient/JschFTPSClient.java]: ok got it, you name your ftp class according the dep impl.

File: ftp/src/main/java/org/talend/components/ftp/source/FTPInputConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+@Documentation("Configuration for FTP source")
+public class FTPInputConfiguration implements Serializable {
+
+    @Option
+    @Documentation("DataSet")
+    private FTPDataSet dataSet;
+
+    @Option
+    @Documentation("Enable debug mode")
+    private boolean debug;
```
Comment (undx) [inline ftp/src/main/java/org/talend/components/ftp/source/FTPInputConfiguration.java]: would have put it in an advanced layout... but as there's nothing in in conf...

File: ftp/src/test/java/org/talend/components/ftp/source/FTPInputTest.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+@Slf4j
+
+@Environment(ContextualEnvironment.class)
+@EnvironmentConfiguration(environment = "Contextual", systemProperties = {})
+
+// @Environment(SparkRunnerEnvironment.class)
+// @EnvironmentConfiguration(environment = "Spark", systemProperties = {
+// @EnvironmentConfiguration.Property(key = "talend.beam.job.runner", value = "org.apache.beam.runners.spark.SparkRunner"),
+// @EnvironmentConfiguration.Property(key = "talend.beam.job.filesToStage", value = ""),
+// @EnvironmentConfiguration.Property(key = "spark.ui.enabled", value = "false")})
```
Comment (undx) [inline ftp/src/test/java/org/talend/components/ftp/source/FTPInputTest.java]: remove if unused

File: ftp/src/test/java/org/talend/components/ftp/source/FTPInputTest.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        configuration.getDataSet().getCsvConfiguration().setFieldSeparator(fieldSeparator);
+        configuration.getDataSet().getCsvConfiguration().setLineConfiguration(new LineConfiguration());
+
+        String configURI = SimpleFactory.configurationByExample().forInstance(configuration).configured().toQueryString();
+
+        try {
+            Job.components().component("input", "FTP://FTPInput?" + configURI).component("output", "test://collector")
+                    .connections().from("input").to("output").build().run();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (undx) [inline ftp/src/test/java/org/talend/components/ftp/source/FTPInputTest.java]: prefer log

File: ftp/src/main/java/org/talend/components/ftp/dataset/FTPDataSet.java (hunk 5 above)
Comment (clesaec) [reply ftp/src/main/java/org/talend/components/ftp/dataset/FTPDataSet.java]: That's a good implementation to retrieve the selected ContentFormat.
An alternative could be 
```
public enum Format {
        CSV(FTPDataSet::getCsvConfiguration, "csv");
        private Function<FTPDataSet, ContentFormat> config;
       ...
}

public ContentFormat getFormatConfiguration() {
      return this.format.config.apply(this);
}
```
So you don't need introspection

Comment (undx) [reply ftp/src/main/java/org/talend/components/ftp/dataset/FTPDataSet.java]: really prefer this one!

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+        UUID uuid = UUID.randomUUID();
+        long l = ByteBuffer.wrap(uuid.toString().getBytes()).getLong();
+        return Long.toString(l, Character.MAX_RADIX);
+    }
+
+    private void checkCurrentStream() {
+        if (currentStream == null
+                || (configuration.getLimitBy().isLimitedByRecords() && currentRecords >= configuration.getRecordsLimit())
+                || (configuration.getLimitBy().isLimitedBySize()
+                        && currentStream.getCurrentSize() >= configuration.getSizeUnit().apply(configuration.getSizeLimit()))) {
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java]: Boom : Null pointer exception pb

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+                try {
+                    recordWriter.end();
+                    recordWriter.flush();
+                    recordWriter.close();
+                    if (ftpClient.isConnected()) {
+                        getFtpClient().disconnect();
+                    }
+                } catch (IOException e) {
+                    log.error(e.getMessage(), e);
+                }
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java]: redondant with release function; may put it in private function

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+            }
+            currentRecords = 0;
+            // Must create new file
+            String path = fileBaseName + "_" + fileIndex++ + "." + configuration.getDataSet().getFormat().getExtension();
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java]: fileIndex never initialized ?

Comment (undx) [reply ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java]: primitve

File: ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+                    e.printStackTrace();
+                }
+
+            }
+            fileIterator = filesToRead.iterator();
+            ContentFormat contentFormat = configuration.getDataSet().getFormatConfiguration();
+            recordReader = recordIORepository.findReader(contentFormat.getClass()).getReader(recordBuilderFactory, contentFormat);
+        }
+
+        if (recordIterator == null || !recordIterator.hasNext()) {
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java]: Use [IteratorComposer.flatMap](https://github.com/Talend/connectors-se/blob/rlecomte/TDI-43870_FTP/common/src/main/java/org/talend/components/common/collections/IteratorComposer.java#L71) 
```
IteratorComposer.of(fileIterator).flatmap( (GenericFTPFile f) -> {
                    final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
                    ...
                    return getFtpClient().retrieveFile(path, buffer);
                })
```
Or even more clearer with a private function that transform GenericFTPFile to a recorid iterator.

File: ftp/src/main/java/org/talend/components/ftp/service/FTPService.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    private FTPClientFactory ftpClientFactory;
+
+    @HealthCheck(ACTION_HEALTH_CHECK)
+    public HealthCheckStatus validateDataStore(@Option final FTPDataStore dataStore) {
+        if (dataStore.getHost() == null || "".equals(dataStore.getHost().trim())) {
+            return new HealthCheckStatus(HealthCheckStatus.Status.KO, i18n.hostRequired());
+        }
+
+        GenericFTPClient ftpClient = getClient(dataStore);
+        try {
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/service/FTPService.java]: use `try (GenericFTPClient ftpClient = getClient(dataStore)) {`, as GenericFTPClient is autocloseable & close function call disconnect

File: ftp/src/main/java/org/talend/components/ftp/dataset/FTPDataSet.java (hunk 5 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/dataset/FTPDataSet.java]: I've been using Java introspection for too long... I'll change to @clesaec solution.

File: ftp/src/main/java/org/talend/components/ftp/service/ftpclient/FTPClientFactory.java (hunk 11 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/service/ftpclient/FTPClientFactory.java]: ftpClient lifecycle is not managed by TCK

File: ftp/src/main/java/org/talend/components/ftp/source/FTPInputConfiguration.java (hunk 13 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/source/FTPInputConfiguration.java]: Yes. What's the point into making an advanced settings if there is no basic settings.

File: ftp/src/test/java/org/talend/components/ftp/source/FTPInputTest.java (hunk 14 above)
Comment (rlecomte-talend) [reply ftp/src/test/java/org/talend/components/ftp/source/FTPInputTest.java]: All Unit test have to be changed (TDI--43936)

File: ftp/pom.xml (hunk 1 above)
Comment (rlecomte-talend) [reply ftp/pom.xml:16]: Actually, ftp won't usetestcontainers. Unit are still to be fixed.

File: ftp/pom.xml (hunk 2 above)
Comment (rlecomte-talend) [reply ftp/pom.xml:31]: ok, done

File: ftp/pom.xml (hunk 4 above)
Comment (rlecomte-talend) [reply ftp/pom.xml:58]: Unit test are still to be fixed. MockFtpServer may disappear from pom.

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java (hunk 17 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java:119]: done

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java (hunk 18 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java:123]: fileIndex is automatically initialized at 0 as a primitive attribute.

File: ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java (hunk 19 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java:102]: done

File: ftp/pom.xml (hunk 3 above)
Comment (rlecomte-talend) [reply ftp/pom.xml:36]: done

File: ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+
+    protected final RecordBuilderFactory recordBuilderFactory;
+
+    protected final I18nMessage i18n;
+
+    protected final RecordIORepository recordIORepository;
+
+    protected List<GenericFTPFile> filesToRead;
+
+    private transient Iterator<GenericFTPFile> fileIterator;
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java:61]: Now that this field is used only in init part of next function, is there an advantage to keep it in class field.
Same question for recordReader field.
(Less field => simpler class :))

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java (hunk 18 above)
Comment (clesaec) [reply ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java:115]: and about puting fileBaseName & fileIndex & shortUUID in a specific class. (more testable ?)

File: ftp/src/main/java/org/talend/components/ftp/service/ftpclient/ApacheFTPClient.java (hunk 10 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/service/ftpclient/ApacheFTPClient.java:41]: Apache impl of FTP

File: ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java [hunk 22]
```diff
@@ -105,18 +110,37 @@ private static String shortUUID() {
...
             // Must create new file
-            String path = fileBaseName + "_" + fileIndex++ + "." + configuration.getDataSet().getFormat().getExtension();
-            log.debug("Creating remote file " + path);
-            currentStream = new SizeAwareOutputStream(getFtpClient().storeFileStream(path));
+            String path = fileBaseName + "_" + fileIndex + "." + configuration.getDataSet().getFormat().getExtension();
+            log.info(i18n.infoCreateRemoteFile(path));
+            OutputStream ftpout = getFtpClient().storeFileStream(path);
+            if (ftpout == null) {
+                log.error(i18n.errorCreateRemoteFile(path, ftpClient.getReplyCode()));
+                checkCurrentStream(nbRetry++);
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/output/FTPOutput.java:135]: Here, why not
```
OutputStream ftpout = getFtpClient().storeFileStream(path);
for (int retry = 1; retry < NB_MAX_RETRIES_PUT && ftpout == null; retry++) {
     log.error(i18n.errorCreateRemoteFile(path, ftpClient.getReplyCode()));
     ftpout = getFtpClient().storeFileStream(path);
}
 if ( ftpout == null) {
      String msg = i18n.errorTooManyRetries(NB_MAX_RETRIES_PUT);
      log.error(msg);
      throw new FTPConnectorException(msg);
   }
```
That's remove recursion and can even be moved in private method (or integrate in GenericFTPClient class)

File: ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java [hunk 23]
```diff
@@ -97,9 +99,14 @@ public Object next() {
             ContentFormat contentFormat = configuration.getDataSet().getFormatConfiguration();
             recordReader = recordIORepository.findReader(contentFormat.getClass()).getReader(recordBuilderFactory, contentFormat);
 
-            recordIterator = IteratorComposer.of(filesToRead.iterator()).map(file -> configuration.getDataSet().getPath()
-                    + (configuration.getDataSet().getPath().endsWith(FTPService.PATH_SEPARATOR) ? "" : FTPService.PATH_SEPARATOR)
-                    + file.getName()).flatmap(path -> {
+            recordIterator = IteratorComposer
+                    .of(filesToRead
+                            .iterator())
+                    .map(file -> configuration.getDataSet().getPath()
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java:105]: With map (or flatmap function), easier to read with explicit type :
`.map( (File file) -> ...`

File: ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java [hunk 24]
```diff
@@ -97,9 +99,14 @@ public Object next() {
...
+                            + (configuration.getDataSet().getPath().endsWith(FTPService.PATH_SEPARATOR) || pathIsFile ? ""
+                                    : FTPService.PATH_SEPARATOR)
+                            + (pathIsFile ? "" : file.getName()))
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/source/FTPInput.java:108]: add a function in FTPService ? (simpler to test and prevent special case (path null for instance) ...)

File: ftp/src/main/java/org/talend/components/ftp/service/FTPService.java [hunk 25]
```diff
@@ -117,10 +117,11 @@ public boolean canWrite(FTPDataSet dataset) {
     public boolean pathIsFile(FTPDataSet dataset) {
         try (GenericFTPClient ftpClient = getClient(dataset.getDatastore())) {
             List<GenericFTPFile> files = ftpClient.listFiles(dataset.getPath());
+            String[] pathElements = dataset.getPath().split(PATH_SEPARATOR);
+            String pathLastElement = pathElements[pathElements.length - 1];
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/service/FTPService.java:121]: Very improbable, but if there are no items in pathElements array ? 
may add `if (pathElements.length == 0) return false` ?

File: ftp/src/main/java/org/talend/components/ftp/service/ftpclient/GenericFTPClient.java [hunk 26]
```diff
@@ -70,4 +71,18 @@ public final void close() {
...
+     * 
+     * @param fullpath
+     * @return
+     */
+    protected final String getFilename(String fullpath) {
+        String[] pathElements = fullpath.split(FTPService.PATH_SEPARATOR);
+        if (pathElements.length > 0) {
+            return pathElements[pathElements.length - 1];
+        }
+        return fullpath;
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/service/ftpclient/GenericFTPClient.java:86]:
```
int pos = fullpath.lastIndexOf(FTPService.PATH_SEPARATOR);
if (pos >=0) {
    return fullpath.substring(pos + 1);
}
return fullpath;
```
is simpler (does not use regexp, nor array)

File: ftp/src/main/java/org/talend/components/ftp/service/FTPService.java [hunk 27]
```diff
@@ -66,7 +74,29 @@ public HealthCheckStatus validateDataStore(@Option final FTPDataStore dataStore)
         }
     }
 
+    @Suggestions(ACTION_SUGGESTION_PATH)
+    public SuggestionValues suggestPath(@Option final FTPDataStore datastore, @Option final String path) {
+        try (GenericFTPClient ftpClient = getClient(datastore)) {
+            System.out.println("Suggestion from " + path);
```
Comment (clesaec) [inline ftp/src/main/java/org/talend/components/ftp/service/FTPService.java:80]: retrieve if it was just for debug or transform to log

Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/service/FTPService.java:80]: removed

File: ftp/src/main/java/org/talend/components/ftp/service/ftpclient/GenericFTPClient.java (hunk 26 above)
Comment (rlecomte-talend) [reply ftp/src/main/java/org/talend/components/ftp/service/ftpclient/GenericFTPClient.java:86]: Not exactly the same. If the path ends with PATH_SEPARATOR, the separator is removed form the name returned.
